certificate.go

```diff
@@ -0,0 +1,133 @@
+}
+
+// FindCertificateOnSlot retrieves previously imported certificate on a specified slot
+//
+// Either (but not all three) of id, label and serial may be nil, in which case they are ignored.
+func FindCertificateOnSlot(slot uint, id []byte, label []byte, serial []byte) (*x509.Certificate, error) {
+	var err error
+	var c *x509.Certificate
+	if err = ensureSessions(instance, slot); err != nil {
+		return nil, err
+	}
+	err = withSession(slot, func(session *PKCS11Session) error {
+		c, err = FindCertificateOnSession(session, slot, id, label, serial)
+		return err
+	})
+	return c, err
+}
+
+// FindCertificateOnSession retrieves a previously imported certificate.
+//
+// Either (but not all three) of id, label and serial may be nil, in which case they are ignored.
+func FindCertificateOnSession(session *PKCS11Session, slot uint, id []byte, label []byte, serial []byte) (*x509.Certificate, error) {
+	var err error
+	var handles []pkcs11.ObjectHandle
+	var template []*pkcs11.Attribute
+
+	if id != nil {
+		template = append(template, pkcs11.NewAttribute(pkcs11.CKA_ID, id))
+	}
+	if label != nil {
+		template = append(template, pkcs11.NewAttribute(pkcs11.CKA_LABEL, label))
+	}
+	if serial != nil {
+		template = append(template, pkcs11.NewAttribute(pkcs11.CKA_SERIAL_NUMBER, serial))
+	}
+	if err = session.Ctx.FindObjectsInit(session.Handle, template); err != nil {
+		return nil, err
+	}
+	defer session.Ctx.FindObjectsFinal(session.Handle)
+	if handles, _, err = session.Ctx.FindObjects(session.Handle, 1); err != nil {
+		return nil, err
+	}
+	if len(handles) == 0 {
+		return nil, ErrCertificateNotFound
+	}
+
+	attributes := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_VALUE, 0),
+	}
+
+	if attributes, err = session.Ctx.GetAttributeValue(session.Handle, handles[0], attributes); err != nil {
+		return nil, err
+	}
+
+	return x509.ParseCertificate(attributes[0].Value)
+}
+
+// ImportCertificate imports a certificate on a specified slot and populates the available attributes
+//
+// The key will have a random label and ID.
+func ImportCertificate(id []byte, label []byte, certificate *x509.Certificate) error {
+	return ImportCertificateOnSlot(instance.slot, id, label, certificate)
+}
+
+// ImportCertificateOnSlot imports a certificate on a specified slot and populates the available attributes
+//
+// Either or both label and/or id can be nil, in which case random values will be generated.
+func ImportCertificateOnSlot(slot uint, id []byte, label []byte, certificate *x509.Certificate) error {
+	var err error
+	if err = ensureSessions(instance, slot); err != nil {
+		return err
+	}
+	err = withSession(slot, func(session *PKCS11Session) error {
+		return ImportCertificateOnSession(session, slot, id, label, certificate)
+	})
+	return err
+}
+
+// ImportCertificateOnSession imports a certificate and populates the available attributes
+//
+// Either or both label and/or id can be nil, in which case random values will be generated.
```

Comment:
> Either or both label and/or id can be nil, in which case random values will be generated.

I cannot see anywhere in the PKCS#11 specification or in the code below where either the label or id have random values generated when they are nil. Is this a PKCS#11 implementation specific observation or have I missed something in the spec? I've been using [v2.20](https://www.cryptsoft.com/pkcs11doc/STANDARD/pkcs-11v2-20.pdf).

If this is an implementation specific behaviour it'd be nice if we normalized and generated values in crypto11 to be consistent across providers.

Reply:
I think you may want to call `generateKeyLabel()`.

Reply:
I made the change to generate the label and id if it is not set.